### PR TITLE
Feature/design controller

### PR DESCRIPTION
**Pull request by andrewtran1995**
**NOTE**: Developed primarily on the Jetson with pair programming, due to the many dependences needed (many of which are installed on the Jetson and readily available).
## Major Changes
* **EventHandler**
   * Complete overhaul of behavior and implementation
   * Behavior defined and influenced by states (4 possible states total)
* **AeroCubeEvent**
   * Ability to encode AeroCubeEvent objects as stringified JSON objects
   * Ability to decode ("rehydrate") AeroCubeEvent objects from JSON objects
* **Controller**
   * All of controller implemented
   * Listens to incoming messages via TCP Server and calls methods depending on signal
* **TcpService**
   * TCP client and server code moved to its own Python module
   * TcpUtil class provides functionality to encode/decode messages for TCP transmission
## Known Issues
* Tests
   * Integration tests for Controller processing a job
   * Integration tests for Event Handler and Controller communicating to process a job
   * Unit tests, especially for Event Handler
* Documentation
   * Event Handler, especially the different states and how state transitions should occur

### CODE REVIEW

File: clearPort5005.sh
Comment (andrewtran1995) [inline clearPort5005.sh]: Should be moved into general-purpose "scripts" folder (along with potentially all other *.sh files currently in the top-level directory right now).

File: controller/controller.py
Comment (andrewtran1995) [inline controller/controller.py]: Params (especially IP and port) could be moved into settings file.

File: controller/controller.py
Comment (andrewtran1995) [inline controller/controller.py:33]: Print statements should be handled by a logger.

File: controller/controller.py
Comment (andrewtran1995) [inline controller/controller.py]: Method created for class demo, and should either be refactored or removed.

File: controller/testController.py
Comment (andrewtran1995) [inline controller/testController.py:5]: Need unit tests for the controller.

Need integration tests for controller and event handler interaction.

File: dataStorage/dataStorage.py
Comment (andrewtran1995) [inline dataStorage/dataStorage.py:19]: Needs unit tests, documentation, and perhaps some refactoring? For example, does not check if image already exists or if valid img path.

File: eventClass/aeroCubeEvent.py
Comment (andrewtran1995) [inline eventClass/aeroCubeEvent.py:75]: Should try to refactor this method to make it more scalable / less brittle. Need to somehow resolve the issue of constructing different types of AeroCubeEvents which may require different arguments in their constructors. Perhaps each class should have a static method that "rehydrates from JSON"?

File: eventClass/aeroCubeSignal.py
Comment (andrewtran1995) [inline eventClass/aeroCubeSignal.py:13]: Perhaps the __str__ behavior could be abstracted somehow -- could be possible if AeroCubeSignal inherits from Enum and everything else inherits from it? Unsure how this behavior could effect Enum behavior.

File: eventClass/aeroCubeSignal.py
Comment (andrewtran1995) [inline eventClass/aeroCubeSignal.py:17]: Need way to signify if a result event is final or not. Could be done with a subclass or a method.

File: eventClass/eventHandler.py
Comment (andrewtran1995) [inline eventClass/eventHandler.py:7]: Class definitely needs more documentation to clarify its behavior (in terms of the different states and how state transitions should occur, both publicly and internally).

File: externalComm/externalComm.py
Comment (andrewtran1995) [inline externalComm/externalComm.py:6]: Could be refactored so that calls from the controller can be more clear.

File: flaskServer/mockTcpEventHandler.py
Comment (andrewtran1995) [inline flaskServer/mockTcpEventHandler.py]: IP and port arguments shoudl be moved to settings file.

File: flaskServer/restEndpoint.py
Comment (andrewtran1995) [inline flaskServer/restEndpoint.py]: IP and port should be moved to settings file.

File: flaskServer/restEndpoint.py
Comment (andrewtran1995) [inline flaskServer/restEndpoint.py]: In order to make the repository cloneable on the Jetson without much issue, perhaps we should try to stick to relative file paths. If so, we should sweep through all of the files (including ImP files) to make the appropriate changes. Something cool could be setting a "root" or "home" directory at the top-level directory of the repo that all Python packages have access to.

**Comment by andrewtran1995** [comment]
@elswenson, would it be appropriate to merge this branch or no?

File: controller/controller.py
Comment (elswenson) [reply controller/controller.py]: Removed seems to be the best choice

File: eventClass/aeroCubeEvent.py
Comment (elswenson) [inline eventClass/aeroCubeEvent.py:63]: Should define a to_json function that returns the object in json representation. __str__ Should not be used for this.

File: eventClass/aeroCubeEvent.py
Comment (elswenson) [reply eventClass/aeroCubeEvent.py:75]: A good start would be to look into better ways of determining the class that the JSON represents, that's where many of the problems lie.

File: eventClass/aeroCubeSignal.py
Comment (elswenson) [reply eventClass/aeroCubeSignal.py:17]: I'm in favor of a subclass

File: eventClass/bundle.py
Comment (elswenson) [inline eventClass/bundle.py:61]: Once again this should be moved to a to_json method, the __str__ method should not be used to create a JSON representation.

File: eventClass/eventHandler.py
Comment (elswenson) [reply eventClass/eventHandler.py:7]: Within the file itself, the functions should be re-ordered, the current order/groupings aren't very coherent.

File: externalComm/commClass.py
Comment (elswenson) [inline externalComm/commClass.py:27]: The API config should be moved to a separate settings file

File: externalComm/commClass.py
Comment (elswenson) [inline externalComm/commClass.py:24]: Need to add logging to this class.

Comment (elswenson) [reply externalComm/commClass.py:24]: We should also look at refactoring the functions in this class into protected/public parts.

File: externalComm/commClass.py
Comment (elswenson) [inline externalComm/commClass.py:84]: Need to check for potential errors in this flow

File: flaskServer/requirements.txt
Comment (elswenson) [inline flaskServer/requirements.txt]: dill should be removed, we are no longer using it.

File: flaskServer/restEndpoint.py
Comment (elswenson) [inline flaskServer/restEndpoint.py:25]: Need to look into handling CORS problems when using a locally hosted web application in a browser to access a local socket process

File: tcpService/tcpUtils.py
Comment (elswenson) [inline tcpService/tcpUtils.py:1]: We should look for other encoding methods

File: dataStorage/dataStorage.py
Comment (elswenson) [inline dataStorage/dataStorage.py:43]: Pickle is not a valid long-term solution for long-term storage, plus reliance on the structure of the classes could cause development problems if the stored files are not cleared when potentially breaking changes are made to class structures.

**Review by elswenson — CHANGES_REQUESTED** [review]
I would rather not merge this until at least some of the problems that have been outlined have been addressed. I'll take a look at this over this weekend to see what can be cleaned up relatively easily.

**Comment by andrewtran1995** [comment]
@elswenson, do you know what other specific changes you would want to see before we merge?

**Comment by elswenson** [comment]
@andrewtran1995 I'll check again today. I'm not entirely sure.

**Pull request merged**
